ledger-api/grpc-definitions/com/daml/ledger/api/v2/event_query_service.proto

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v2;
+
+import "com/daml/ledger/api/v1/event.proto";
+import "com/daml/ledger/api/v1/event_query_service.proto";
+
+option java_outer_classname = "EventQueryServiceOuterClass";
+option java_package = "com.daml.ledger.api.v2";
+option csharp_namespace = "Com.Daml.Ledger.Api.V2";
+
+// Query events by contract id or key.
+service EventQueryService {
+
+  // Get the create and the consuming exercise event for the contract with the provided ID.
+  // No events will be returned for contracts that have been pruned because they 
+  // have already been archived before the latest pruning offset. 
+  rpc GetEventsByContractId (com.daml.ledger.api.v1.GetEventsByContractIdRequest) returns (GetEventsByContractIdResponse);
+
+}
+
+message GetEventsByContractIdResponse {
+
+  // The create event for the contract with the ``contract_id`` given in the request
+  // provided it exists and has not yet been pruned.
+  // Optional
+  Created created = 1;
+
+  // The archive event for the contract with the ``contract_id`` given in the request
+  // provided such an archive event exists and it has not yet been pruned.
+  // Optional
+  Archived archived = 2;
+
+}
+
+message Created {
+
+  // Required
+  com.daml.ledger.api.v1.CreatedEvent created_event = 1;
+
+  // The domain ID where the contract was created
+  // Required
+  string domain_id = 2;
+
+}
+
+message Archived {
+
+  // Required
```

Comment:
Is it really a problem? I think it is a corner case (domain drift is bigger than the lifespan of a contract, and we are querying earlier than the creation is witnessed on the participant) to start with, and not sure this would propose a problem for all usages.
On the other hand implementing this closely to what this endpoint meant before (so stricly creation and archival): would mean lot of creation missing for multi-hosted parties. (please note this is a participant local endpoint).
I share your idea with this endpoint, that it should be sequence of all events associated with a contract (similarly how it looked for keys before).
I am not entirely certain that this is the right time to tackle this. Let's discuss this.

Reply:
@cocreature when would such a case occur ? If the stakeholder is single-hosted, then if an archive is emitted on domain D, you should see the activation (create or transfer-in) before, no ?
On the other hand, it could occur for multi-hosted parties if the topology change between the activation and the archive but then shouldn't we expect some kind of party migration (and thus, some activation event)?

Reply:
@rgugliel-da you're right that you will see an activation event before but the current endpoint doesn't show you activation events, it shows you CreatedEvents but not the transfer in. So in that case you can see a transfer in (not emitted here) and an archive (emitted here).

We don't use this endpoint so definitely not a blocker for us. Happy to postpone the discussion but I think we should make sure before 3.0 (but doesn't have to be before June) that we at least have an idea for how to extend it in a non-breaking way later.

Reply:
As discussed we keep this as it is now with the aforementioned impediments, and we may improve for 3.0